### ISSUE TITLE
Failed to generate Il2Cpp interop assemblies

### ISSUE THREAD

**Issue by gsdaemon** [BepInEx/BepInEx#479]
### What happened?

Using the bleeding edge version of BepInEx with il2cpp obfuscated game.
After installing the BepInEx and launching the gave, the console start with the output from the logs provided, then it exits.

### BepInEx Distribution

Bleeding Edge from BepisBuilds

### Log outputs

```txt
https://pastebin.com/HFrE4xJA
```

### Environment

```markdown
- OS: Windows 10
- BepInEx: 6.0.0-be.652+8621f84
- Game: Evony TKR
```

**Comment by gsdaemon** [comment]
Possible way to reproduce:
- install the game after downloading it from their official website: https://pc.evony.com/pc/downloada.html
- open the game once to make sure it works
- install BepInEx
- start the game again

**Comment by ghorsington** [comment]
Obfuscated games generally also come with some Il2Cpp metadata tampering to make work of Cpp2IL and Il2CppInterop hard/impossible. We can't support such cases out of the box because obfuscation is generally a per-game issue. You'll have to likely solve any tampering protections and obfuscation before those tools can be run (and BepInEx launched).

There are plans to make a "freestanding" mode for BepInEx where Cpp2IL nor Il2CppInterop are loaded in which case you'll be able to load in BepInEx at the cost of not being able to use either tool for easier modding. However, it's a lower priority now.

**Comment by gsdaemon** [comment]
Hi @ghorsington 

Thank you for your answer. I tried again on a fresh install of Evony using BepInEx 6.0.0-pre.1.
This time only the LogOutput.log file was generated, without ErrorLog.log, and the errors at the end are different this time. Not sure if the reason is still the same.

The full content of LogOutput.log:
https://pastebin.com/VU3bHURR

Thank you!

**Comment by ghorsington** [comment]
A minor update: Il2CppInterop was updated to address a special case where the error you reported might be caused. Please give the latest [Bleeding Edge](https://builds.bepinex.dev/projects/bepinex_be) build (6.0.0-be.653) a try.

**Comment by gsdaemon** [comment]
Hi @ghorsington . Thank you for consideration of my issue!
Just tested it out, but unfortunately the error is still there.
Log output below:

https://pastebin.com/dtumNeUA

Best regards

**Comment by ghorsington** [comment]
Looks like it indeed fixed the `Sequence contains no elements` issue as was intended. The xref issue you're having in the logs is related to the game assembly obfuscation. As I mentioned before, fixing that up is pretty much up to you.

**Comment by ghorsington** [comment]
After testing, I could not replicate any of the errors presented in the logs. However, I noticed some other errors that are now fixed.

You can try the latest BE build: [BepInEx-Unity.IL2CPP-win-x86-6.0.0-be.654+c0b9eae.zip](https://builds.bepinex.dev/projects/bepinex_be/654/BepInEx-Unity.IL2CPP-win-x86-6.0.0-be.654%2Bc0b9eae.zip)

I verified that it indeed works fine with the game:
![image](https://user-images.githubusercontent.com/13762037/189498513-f8247630-4ef7-4f7b-86af-c85dcac14512.png)

If you decide to use latest BE, make sure to first delete any other BepInEx folders and files you have before installing.

**Comment by gsdaemon** [comment]
Wow!
Worked like a charm this time up to the point from your screenshot, including the error about `GetBattlePassReply`.
But then another error appeared right when the entry screen of the game appears and each time when you get back to it:
![image](https://user-images.githubusercontent.com/6863250/189502136-e6b095ea-8118-46f4-8a3b-8eaae193bb1c.png)

```
[Error  :     Unity]   at TaskManager.ExistMileStoneReward () [0x00000] in <filename unknown>:0
  at UIBuildingAltar.UpdateBuildingMark (.UIBuildingCompleteMark currentMark, .BuildingBaseTimeInfo timeInfo) [0x00000] in <filename unknown>:0
  at UIBuildingBase.UpdateUIByTime (.BuildingBaseTimeInfo timeInfo) [0x00000] in <filename unknown>:0
  at UIBuildingBase.OnUpdateUIByTime (DataChangeType type, System.Object[] param) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at OnScrollStatusChanged.Invoke (TableStatus oldStatus, TableStatus newStatus) [0x00000] in <filename unknown>:0
  at DataChangeBroadcast.RaiseEvent (DataChangeType type, System.Object[] objs) [0x00000] in <filename unknown>:0
  at BuildingManager.OnTimer (.Timer timer) [0x00000] in <filename unknown>:0
  at Frankfort.Threading.ThreadPoolSchedulerEvent.Invoke (Frankfort.Threading.IThreadWorkerObject[] finishedObjects) [0x00000] in <filename unknown>:0
  at Timer.Trigger () [0x00000] in <filename unknown>:0
  at TimerManager.FixedUpdate () [0x00000] in <filename unknown>:0
```

Not sure how big of an issue are these.

Beside this, is there any plugin which should work with this latest build?
When trying UnityExplorer is will simply ignore it and the console will say `0 plugins to load`, and when trying BepInExConfigManager, the following error is shown and, after the alert popup "BepInEx failed to load", it will exit.

```
System.IO.FileLoadException: Could not load file or assembly 'BepInExConfigManager.Il2Cpp.Patcher, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromPath(IntPtr ptrNativeAssemblyLoadContext, String ilPath, String niPath, ObjectHandleOnStack retAssembly)
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at BepInEx.Preloader.Core.Patching.AssemblyPatcher.AddPatchersFromDirectory(String directory) in /home/runner/work/BepInEx/BepInEx/BepInEx.Preloader.Core/Patching/AssemblyPatcher.cs:line 150
   at BepInEx.Unity.IL2CPP.Preloader.Run() in /home/runner/work/BepInEx/BepInEx/Runtimes/Unity/BepInEx.Unity.IL2CPP/Preloader.cs:line 70
   at BepInEx.Unity.IL2CPP.UnityPreloaderRunner.PreloaderMain() in /home/runner/work/BepInEx/BepInEx/Runtimes/Unity/BepInEx.Unity.IL2CPP/UnityPreloadRunner.cs:line 29
   at Doorstop.Entrypoint.Start() in /home/runner/work/BepInEx/BepInEx/Runtimes/Unity/BepInEx.Unity.IL2CPP/DoorstopEntrypoint.cs:line 39
```

**Comment by ghorsington** [comment]
>  Please correct me if I'm wrong (coming from web developing and still very new into C#

You're correct about both points. UnityExplorer hasn't yet been ported to support the latest BE builds yet.

> Is there a good source or example of how to access/patch within the code the il2cpp classes and methods of the game itself?  Perhaps with code completion if at all possible. 

The general approach is to use [Harmony](https://harmony.pardeike.net/) to write patches at runtime. With Il2Cpp games, you don't get access to the game code directly but have to resort to using a proper RE tool like Ghidra or IDA. You can, however, look at the names of the classes and game methods a bit more easily by using dnSpy and opening the interop assemblies (assemblies inside the `BepInEx/interop` folder). Interop assemblies have no game code, but all the classes, methods and fields are intact.

For further ideas/help, I suggest joining [BepInEx Discord](https://discord.gg/MpFEDAg). There are a lot of logs with examples of people using Harmony in Il2Cpp games.

**Comment by ghorsington** [comment]
Closing this issue as the main issue has been resolved via c0b9eae55804aab85235308d715cf63eebc9cfb0 and 44cb650e6394880ddcb98c340617e962b4edfac8

**Comment by krulci** [comment]
Have you got UnityExplorer to work with Evony?
I find Evony use such an ancient Unity build and the interop does not generate 'Boolean Il2CppSystem.Type.op_Inequality(Il2CppSystem.Type, Il2CppSystem.Type)'
